### PR TITLE
socket: add SO_KEEPALIVE option for upstream connections.

### PR DESCRIPTION
**Pull request by JonathanO**
*Description*:
Adds TCP Keepalive support for upstream connections. This can be configured on the cluster manager level, and overridden on the cluster level.

*Risk Level*: Medium

*Testing*:
Unit tests have been added. It appears to run and work.

*Docs Changes*:
envoyproxy/data-plane-api#614

Fixes envoyproxy/envoy#3028

*API Changes*:
envoyproxy/data-plane-api#614

### CODE REVIEW

File: source/common/upstream/upstream_impl.cc [hunk 1]
```diff
@@ -102,6 +109,21 @@ HostImpl::createConnection(Event::Dispatcher& dispatcher, const ClusterInfo& clu
...
   }
+  if (cluster.features() & ClusterInfo::Features::USE_TCP_KEEPALIVE) {
+    if (!cluster_options) {
+      cluster_options = std::make_shared<Network::ConnectionSocket::Options>();
+      if (options) {
+        *cluster_options = *options;
+      }
+    }
+    cluster_options->emplace_back(new Network::TcpKeepaliveOptionImpl(
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(cluster.tcpKeepaliveSettings(), keepalive_probes,
```
Comment (JonathanO) [inline source/common/upstream/upstream_impl.cc]: This is horrible and I shouldn't have done it this way. I'll move it out to a struct that'll be populated at construction time.

File: source/common/network/tcp_keepalive_option_impl.cc [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#include "common/network/tcp_keepalive_option_impl.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+bool TcpKeepaliveOptionImpl::setOption(Network::Socket& socket,
+                                       Network::Socket::SocketState state) const {
+  if (state == Socket::SocketState::PreBind) {
```
Comment (JonathanO) [inline source/common/network/tcp_keepalive_option_impl.cc]: Will add a new state, perhaps PreConnect, since PreBind is a bit too vague for this option.

File: include/envoy/upstream/cluster_manager.h [hunk 3]
```diff
@@ -150,6 +150,12 @@ class ClusterManager {
    */
   virtual const envoy::api::v2::core::BindConfig& bindConfig() const PURE;
 
+  /**
+   * @return const envoy::api::v2::UpstreamConnectionOptions& cluster manager wide connection
+   * options for new upstream connections.
+   */
+  virtual const envoy::api::v2::UpstreamConnectionOptions& connectionOptions() const PURE;
```
Comment (ggreenway) [inline include/envoy/upstream/cluster_manager.h]: Given that configs are usually machine-generated, should we remove this (and the corresponding data-plane-api field) and just set this on each cluster?

File: source/common/upstream/upstream_impl.cc [hunk 4]
```diff
@@ -67,9 +69,38 @@ uint64_t parseFeatures(const envoy::api::v2::Cluster& config,
...
+                        const envoy::api::v2::UpstreamConnectionOptions connection_options) {
+  if (connection_options.has_tcp_keepalive()) {
+    return Network::TcpKeepaliveConfig{
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(connection_options.tcp_keepalive(), keepalive_probes,
+                                        absl::optional<uint32_t>()),
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(connection_options.tcp_keepalive(), keepalive_time,
+                                        absl::optional<uint32_t>()),
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(connection_options.tcp_keepalive(), keepalive_interval,
+                                        absl::optional<uint32_t>())};
+  } else if (config.upstream_connection_options().has_tcp_keepalive()) {
```
Comment (ggreenway) [inline source/common/upstream/upstream_impl.cc]: These two cases can be refactored and collapsed into one, just operating on a different object.

File: source/common/upstream/upstream_impl.cc [hunk 5]
```diff
@@ -102,6 +133,16 @@ HostImpl::createConnection(Event::Dispatcher& dispatcher, const ClusterInfo& clu
   } else {
     cluster_options = options;
   }
+  if (cluster.features() & ClusterInfo::Features::USE_TCP_KEEPALIVE) {
+    if (!cluster_options) {
```
Comment (ggreenway) [inline source/common/upstream/upstream_impl.cc]: Please clean up with duplicate logic from the FREEBIND case above. This is hard to follow.

File: source/common/upstream/upstream_impl.cc [hunk 6]
```diff
@@ -102,6 +133,16 @@ HostImpl::createConnection(Event::Dispatcher& dispatcher, const ClusterInfo& clu
   } else {
     cluster_options = options;
   }
+  if (cluster.features() & ClusterInfo::Features::USE_TCP_KEEPALIVE) {
```
Comment (ggreenway) [inline source/common/upstream/upstream_impl.cc]: Instead of adding a big to features(), can you just condition on whether cluster.tcpKeepaliveSettings() is set or not, and turn it into an optional?

File: test/common/network/tcp_keepalive_option_impl_test.cc [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+  NiceMock<MockListenSocket> socket_;
+  Api::MockOsSysCalls os_sys_calls_;
+  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls{&os_sys_calls_};
+};
+
+// We fail to set the option when the underlying setsockopt syscall fails.
+TEST_F(TcpKeepaliveOptionImplTest, SetOptionTcpKeepaliveFailure) {
+  TcpKeepaliveOptionImpl socket_option{{}};
+#ifdef SO_KEEPALIVE
```
Comment (ggreenway) [inline test/common/network/tcp_keepalive_option_impl_test.cc]: if (ENVOY_SOCKET_SO_KEEPALIVE.has_value())

File: source/common/network/tcp_keepalive_option_impl.cc [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    return setTcpKeepalive(socket, keepalive_config_.keepalive_probes_,
+                           keepalive_config_.keepalive_time_,
+                           keepalive_config_.keepalive_interval_);
+  }
+  return true;
+}
+
+bool TcpKeepaliveOptionImpl::setTcpKeepalive(Socket& socket,
+                                             const absl::optional<uint32_t>& keepalive_probes,
```
Comment (ggreenway) [inline source/common/network/tcp_keepalive_option_impl.cc]: These 3 params can be passed by value instead of by const&

File: source/common/network/tcp_keepalive_option_impl.cc [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+  const int optOne = 1;
+  const socklen_t optOneLen = sizeof(optOne);
+  int error = os_syscalls.setsockopt(socket.fd(), SOL_SOCKET, ENVOY_SOCKET_SO_KEEPALIVE.value(),
+                                     &optOne, optOneLen);
+  if (error != 0) {
+    ENVOY_LOG(warn, "Setting SO_KEEPALIVE on socket failed: {}", strerror(error));
+    return false;
+  }
+  if (keepalive_probes.has_value()) {
```
Comment (ggreenway) [inline source/common/network/tcp_keepalive_option_impl.cc]: Please factor out a function for the pattern "if option doesn't exist then return false; set the option; check the error" and call it 4 times.

File: include/envoy/upstream/cluster_manager.h (hunk 3 above)
Comment (JonathanO) [reply include/envoy/upstream/cluster_manager.h]: This allows it to be set in the bootstrap config easily, and makes it's behaviour consistent with upstream_bind_options.

File: source/common/upstream/upstream_impl.cc (hunk 6 above)
Comment (JonathanO) [reply source/common/upstream/upstream_impl.cc]: I copied this pattern from the HTTP2 feature, I assumed there was maybe a performance reason for using the features flags rather than checking an optional field. If there's not, then I'll change it.

Comment (JonathanO) [reply source/common/upstream/upstream_impl.cc]: Looking at this further, I'm wondering if I should actually refactor to instantiate the TcpKeepaliveOptionImpl (and also UpstreamSocketOption for freebind) at ClusterInfoImpl construct time and store them both in a Options vector on ClusterInfo.
That removes them both from Features.
That should also greatly simplify the code in createConnection, since there's no need to have different cases for them (it's just a case of merging options vectors together is necessary.)
This should be safe as neither Option implementation stores any state that differs between connections to the same cluster.

What do you think?

Comment (ggreenway) [reply source/common/upstream/upstream_impl.cc]: My best guess for why this is the way it is: the features flags came befofe the current iteration of socket-option-setters.

I think your suggested refactoring sounds great, and should be a nice cleanup. @htuch do you know any more history on this, or any objections to this cleanup?

Comment (JonathanO) [reply source/common/upstream/upstream_impl.cc]: I've made an attempt at the refactor (based off this branch) in this commit https://github.com/JonathanO/envoy/commit/bf4bf78b3739529f19d899e5dfabdca9ee1f6196 which I think looks a bit cleaner.

I must admit that I'm (re)learning modern C++ by trial and error, and I have no idea if the shared pointer usage in that commit is right or even necessary.
If it looks sane then I'll merge it into this PR.

File: source/common/network/tcp_keepalive_option_impl.h [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+#include "common/common/logger.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Network {
+
+// Optional variant of setsockopt(2) optname. The idea here is that if the option is not supported
+// on a platform, we can make this the empty value. This allows us to avoid proliferation of #ifdef.
+typedef absl::optional<int> SocketOptionName;
```
Comment (ggreenway) [inline source/common/network/tcp_keepalive_option_impl.h]: This should go in a common header somewhere, instead of copy/pasting it.

I think it might be nice to put all of the socket option preprocessor checking into a single header file. Or better yet, put the declarations in a header, and the definitions in a .cc so all the system headers for the constants don't need to be included in large parts of the codebase. Thoughts on that idea @htuch ?

File: source/common/upstream/upstream_impl.cc [hunk 11]
```diff
@@ -54,33 +55,51 @@ getSourceAddress(const envoy::api::v2::Cluster& cluster,
...
+  // Cluster IP_FREEBIND settings, when set, will override the cluster manager wide settings.
+  if ((bind_config.freebind().value() && !config.upstream_bind_config().has_freebind()) ||
+      config.upstream_bind_config().freebind().value()) {
+    cluster_options->emplace_back(new UpstreamSocketOption(true));
+  }
+  if (config.upstream_connection_options().has_tcp_keepalive()) {
+    cluster_options->emplace_back(
+        new Network::TcpKeepaliveOptionImpl(parseTcpKeepaliveConfig(config)));
+  }
+  return cluster_options;
```
Comment (ggreenway) [inline source/common/upstream/upstream_impl.cc]: if cluster_options is empty, return nullptr?

File: source/common/upstream/upstream_impl.cc [hunk 12]
```diff
@@ -54,33 +55,51 @@ getSourceAddress(const envoy::api::v2::Cluster& cluster,
...
+
+const Network::ConnectionSocket::OptionsSharedPtr
+parseClusterSocketOptions(const envoy::api::v2::Cluster& config,
+                          const envoy::api::v2::core::BindConfig bind_config) {
+  Network::ConnectionSocket::OptionsSharedPtr cluster_options =
+      std::make_shared<Network::ConnectionSocket::Options>();
+  // Cluster IP_FREEBIND settings, when set, will override the cluster manager wide settings.
+  if ((bind_config.freebind().value() && !config.upstream_bind_config().has_freebind()) ||
+      config.upstream_bind_config().freebind().value()) {
+    cluster_options->emplace_back(new UpstreamSocketOption(true));
```
Comment (ggreenway) [inline source/common/upstream/upstream_impl.cc]: std::make_shared

File: source/common/upstream/upstream_impl.cc [hunk 13]
```diff
@@ -54,33 +55,51 @@ getSourceAddress(const envoy::api::v2::Cluster& cluster,
...
+  }
+  if (config.upstream_connection_options().has_tcp_keepalive()) {
+    cluster_options->emplace_back(
+        new Network::TcpKeepaliveOptionImpl(parseTcpKeepaliveConfig(config)));
```
Comment (ggreenway) [inline source/common/upstream/upstream_impl.cc]: std::make_shared

File: source/common/upstream/upstream_impl.cc [hunk 14]
```diff
@@ -99,19 +118,22 @@ Network::ClientConnectionPtr
 HostImpl::createConnection(Event::Dispatcher& dispatcher, const ClusterInfo& cluster,
                            Network::Address::InstanceConstSharedPtr address,
                            const Network::ConnectionSocket::OptionsSharedPtr& options) {
-  Network::ConnectionSocket::OptionsSharedPtr cluster_options;
-  if (cluster.features() & ClusterInfo::Features::FREEBIND) {
-    cluster_options = std::make_shared<Network::ConnectionSocket::Options>();
+  Network::ConnectionSocket::OptionsSharedPtr connection_options;
+  if (cluster.clusterSocketOptions() && !cluster.clusterSocketOptions().get()->empty()) {
```
Comment (ggreenway) [inline source/common/upstream/upstream_impl.cc]: if (cluster.clusterSocketOptions() != nullptr && !cluster.clusterSocketOptions()->empty()) {

File: source/common/upstream/upstream_impl.cc [hunk 15]
```diff
@@ -99,19 +118,22 @@ Network::ClientConnectionPtr
...
     if (options) {
-      *cluster_options = *options;
+      connection_options = std::make_shared<Network::ConnectionSocket::Options>();
+      *connection_options = *options;
+      copy(cluster.clusterSocketOptions().get()->begin(),
```
Comment (ggreenway) [inline source/common/upstream/upstream_impl.cc]: std::copy and std::back_inserter?

File: source/common/network/tcp_keepalive_option_impl.cc [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+  }
+  return true;
+}
+
+bool TcpKeepaliveOptionImpl::setTcpKeepalive(Socket& socket,
+                                             absl::optional<uint32_t> keepalive_probes,
+                                             absl::optional<uint32_t> keepalive_time,
+                                             absl::optional<uint32_t> keepalive_interval) {
+  int error;
+  error = setSocketOption(socket, SOL_SOCKET, ENVOY_SOCKET_SO_KEEPALIVE, 1);
```
[the remark this answers is not in the document]
Comment (JonathanO) [reply source/common/network/tcp_keepalive_option_impl.cc]: Under Linux there's no way to turn it on at the OS level, only tweak the parameters. It appears that under FreeBSD and OSX it's possible to turn it on for everyone with "net.inet.tcp.always_keepalive", but I don't know if it's possible to explicitly disable it using SO_KEEPALIVE=0 in that case.

File: source/common/network/tcp_keepalive_option_impl.h (hunk 10 above)
Comment (JonathanO) [reply source/common/network/tcp_keepalive_option_impl.h]: I was considering making this an absl::optional<pair<int, int>> so that it can contain the "level" as well as "optname" value. I think I'd definitely want to do that if I make these a single set of definitions, which I agree might be cleaner.

File: source/common/network/tcp_keepalive_option_impl.h [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+#ifdef TCP_KEEPCNT
+#define ENVOY_SOCKET_TCP_KEEPCNT Network::SocketOptionName(TCP_KEEPCNT)
+#else
+#define ENVOY_SOCKET_TCP_KEEPCNT Network::SocketOptionName()
+#endif
+
+#ifdef TCP_KEEPIDLE
+#define ENVOY_SOCKET_TCP_KEEPIDLE Network::SocketOptionName(TCP_KEEPIDLE)
+#elif TCP_KEEPALIVE // OS X I think
```
[the remark this answers is not in the document]
Comment (JonathanO) [reply source/common/network/tcp_keepalive_option_impl.h]: OK. Do i need to support OSes other than OSX and Linux? I think there's at least one where the options are in ms rather than s, and Windows seems to use a single option with a struct, which would make things a little messier.

[the remark this answers is not in the document]
Comment (ggreenway) [reply source/common/network/tcp_keepalive_option_impl.h]: Linux is required. Macos is optional. I don't think envoy builds on any other platforms.

File: source/common/upstream/upstream_impl.cc (hunk 14 above)
Comment (JonathanO) [reply source/common/upstream/upstream_impl.cc]: Can be simplified to just the nullptr check if parseClusterSocketOptions() returns nullptr if the options are empty, as you suggest.

File: api/envoy/api/v2/core/address.proto [hunk 18]
```diff
@@ -58,6 +58,20 @@ message SocketAddress {
   bool ipv4_compat = 6;
 }
 
+message TcpKeepalive {
```
[the remark this answers is not in the document]
Comment (JonathanO) [reply api/envoy/api/v2/core/address.proto]: I swear I replied to this, but the comment seems to be missing now.
The reason for the single special case is that MacOS uses a different name for that one constant (TCP_KEEPIDLE vs TCP_KEEPALIVE.) The other constants all have the same name as for Linux. All of these settings are valid for both Linux and MacOS.

File: source/common/network/tcp_keepalive_option_impl.cc [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+int TcpKeepaliveOptionImpl::setSocketOption(Socket& socket, int level,
+                                            Network::SocketOptionName optname, int value) {
+  if (!optname.has_value()) {
+    errno = ENOTSUP;
+    return -1;
+  }
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  return os_syscalls.setsockopt(socket.fd(), level, optname.value(), &value, sizeof(value));
```
[the remark this answers is not in the document]
Comment (JonathanO) [reply source/common/network/tcp_keepalive_option_impl.cc]: I'm going to change SocketOptionName to be absl::optional<std::pair<int, int>>, so it can contain the "level" as well as "optname." That should make it possible to clean up the socket option code a bit and maybe make a common Socket::Option implementation, which takes a SocketOptionName and value as constructor args, possible. That might then allow TCP Keepalive, and other simple options, to be implemented with as a factory that produces a vector of options.

SocketOptionImpl has special handling for checking if the socket it's being applied to is IP, and setting different options for IPV6 vs 4. I think that's going to have to remain slightly special (although it might be able to use the generic implementation under the hood.)

[the remark this answers is not in the document]
Comment (htuch) [reply source/common/network/tcp_keepalive_option_impl.cc]: Awesome. @bmetzdorf for visibility, so that you folks don't collide on this. I could merge #2793 as is if that works for you both.

[the remark this answers is not in the document]
Comment (bmetzdorf) [reply source/common/network/tcp_keepalive_option_impl.cc]: SGTM, I can address the open aesthetic items from #2793 in another PR.

[the remark this answers is not in the document]
Comment (htuch) [reply source/common/network/tcp_keepalive_option_impl.cc]: OK, will merge.

[the remark this answers is not in the document]
Comment (htuch) [reply source/common/network/tcp_keepalive_option_impl.cc]: https://github.com/envoyproxy/envoy/pull/2793 merged.

File: source/common/network/addr_family_aware_socket_impl.cc [hunk 20]
```diff
@@ -0,0 +1,58 @@
+#include "common/network/addr_family_aware_socket_impl.h"
```
Comment (htuch) [inline source/common/network/addr_family_aware_socket_impl.cc]: This file is missing a bunch of code coverage.

File: source/common/network/socket_option_factory.cc [hunk 21]
```diff
@@ -0,0 +1,59 @@
+#include "common/network/socket_option_factory.h"
```
Comment (htuch) [inline source/common/network/socket_option_factory.cc]: This file is missing a bunch of code coverage.

File: source/common/network/socket_option_impl.h [hunk 22]
```diff
@@ -15,69 +16,115 @@ namespace Network {
...
+
+class SocketOptionsImpl : public Socket::Option, Logger::Loggable<Logger::Id::connection> {
+public:
+  SocketOptionsImpl() {}
+
+  void addOption(std::unique_ptr<Socket::Option> option) { options_.push_back(std::move(option)); }
+
+  // Socket::Option
+  bool setOption(Socket& socket, Socket::SocketState state) const override {
+    bool good = true;
```
Comment (htuch) [inline source/common/network/socket_option_impl.h]: Nit: prefer renaming this to be something more descriptive, e.g. `all_socket_options_applied`. Coming to think of it, I'm wondering exactly what semantics we want here.. do we want to abort early and just `return false` in the inner loop rather than continue to apply options?

File: source/common/network/socket_option_impl.h [hunk 23]
```diff
@@ -15,69 +16,115 @@ namespace Network {
...
+
+  static int setSocketOption(Socket& socket, Network::SocketOptionName optname, int value);
+
+private:
+  const Socket::SocketState in_state_;
+  const Network::SocketOptionName optname_;
+  const int value_;
+};
+
+class SocketOptionsImpl : public Socket::Option, Logger::Loggable<Logger::Id::connection> {
```
Comment (htuch) [inline source/common/network/socket_option_impl.h]: Why do we need this given that `Socket` already has a collection of options? Doesn't this then become a collection of options for a collection of options?

Comment (ggreenway) [reply source/common/network/socket_option_impl.h]: The two ideas I like are the last one (always use a special collection object) or for the factory to always return a vector/list (Socket::Options). I lean towards always using Socket::Options because then there isn't an issue with weird nesting.

With either of these approaches, I think eventually it would be better for each socketoption to have it's own object. You don't have to make that change as part of this PR if you don't want, but having 1 option per object, and a consistent way to dealing with a collection/set of these objects, would make this a bit cleaner.

Comment (htuch) [reply source/common/network/socket_option_impl.h]: Yeah, I also like:
* One socket option per object.
* Factory returns always `Socket::Options`

Comment (JonathanO) [reply source/common/network/socket_option_impl.h]: OK, I'll go that path, and for this PR I'll add a bunch of shared utility functions for managing these collections to avoid duplicating code in listener and upstream (rather than adding a new collection type with methods for the needed operations.)

Comment (JonathanO) [reply source/common/network/socket_option_impl.h]: I've just pushed the change to get rid of SocketOptionsImpl. It would definitely look even cleaner with a new collection type, rather than static utility methods I added on Socket, but I think maybe this is already a sufficient improvement? I'm still missing test coverage as @htuch pointed out, which I'll fix tomorrow.

File: include/envoy/network/listen_socket.h [hunk 24]
```diff
@@ -6,6 +6,8 @@
 #include "envoy/common/pure.h"
 #include "envoy/network/address.h"
 
+#include "absl/types/optional.h"
```
Comment (JonathanO) [inline include/envoy/network/listen_socket.h]: Unneeded include

**Comment by ggreenway** [comment]
@JonathanO You have merge conflicts to resolve.
